qt/src/function.rs

```diff
@@ -492,130 +498,218 @@ pub fn read_directory(arg: &Variant) -> Variant {
+                return Ok(FileDetails{
+                    state: FileState::BadFile,
                     format: None,
                     info: vec![],
                 })
-            } else {
-                let ro = meta.permissions().readonly();
-                let state = if ro { "readable" } else { "writable" }.to_string();
-                let peek = goblin::peek(&mut fd);
-                debug!("peek: {:?}", &peek);
-                // TODO: make this prettier, a big hack right now
-                if peek.is_err() {
-                    return Ok(FileDetails{
-                        state: "bad file".to_string(),
-                        format: None,
-                        info: vec![],
-                    })
-                }
-                match peek.unwrap() {
-                    Hint::Elf(data) => {
-                        let endianness = if data.is_lsb { "LittleEndian" } else { "BigEndian" };
-                        let class =
-                            if let Some(is_64) = data.is_64 {
-                                if is_64  {
-                                    "ELF64"
-                                } else {
-                                    "ELF32"
-                                }
+            }
+            match peek.unwrap() {
+                Hint::Elf(data) => {
+                    let endianness = if data.is_lsb { "LittleEndian" } else { "BigEndian" };
+                    let class =
+                        if let Some(is_64) = data.is_64 {
+                            if is_64  {
+                                "ELF64"
                             } else {
-                                "UNKNOWN"
-                            };
-                        Ok(FileDetails{
-                            state: state,
-                            format: Some("elf".to_string()),
-                            info: vec![format!("{:?}, {:?}", class, endianness)]
-                        })
-                    },
-                    Hint::PE => {
-                        Ok(FileDetails{
-                            state: state,
-                            format: Some("pe".to_string()),
-                            info: vec!["PE".to_string()],
-                        })
-                    },
-                    Hint::Mach => {
+                                "ELF32"
+                            }
+                        } else {
+                            "UNKNOWN"
+                        };
+                    Ok(FileDetails{
+                        state: state,
+                        format: Some(FileFormat::ELF),
+                        info: vec![format!("{:?}, {:?}", class, endianness)]
+                    })
+                },
+                Hint::PE => {
+                    Ok(FileDetails{
+                        state: state,
+                        format: Some(FileFormat::PE),
+                        info: vec!["PE".to_string()],
```

Comment:
Why are you using strings in those sections? Didn't you already write a `ToString` implementation for all enums?